internal/cli/demo/k3d.go

```diff
@@ -0,0 +1,165 @@
+	var clusterListJson bytes.Buffer
+	stderr := logger.Get(ctx).Writer(logger.WarnLvl)
+	if err := k.command(ctx, cmd, &clusterListJson, stderr, true); err != nil {
+		return nil, err
+	}
+
+	var clusters []cluster
+	if err := json.Unmarshal(clusterListJson.Bytes(), &clusters); err != nil {
+		return nil, fmt.Errorf("invalid JSON output from cluster list: %v", err)
+	}
+
+	clusterNames := make([]string, len(clusters))
+	for i := range clusters {
+		clusterNames[i] = clusters[i].Name
+	}
+	return clusterNames, nil
+}
+
+func (k *K3dClient) CreateCluster(ctx context.Context, clusterName string) error {
+	cmd := []string{
+		"cluster",
+		"create", clusterName,
+		"--registry-create",
+		"--kubeconfig-switch-context",
+		"--kubeconfig-update-default",
+		"--no-hostip",
+		"--no-image-volume",
+		"--no-lb",
+		"--label", fmt.Sprintf("%s@%s", docker.BuiltByTiltLabelStr, "server[0]"),
+	}
+	stdout := logger.Get(ctx).Writer(logger.DebugLvl)
+	stderr := logger.Get(ctx).Writer(logger.WarnLvl)
+	if err := k.command(ctx, cmd, stdout, stderr, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *K3dClient) DeleteCluster(ctx context.Context, clusterName string, wait bool) error {
+	cmd := []string{
+		"cluster",
+		"delete", clusterName,
+	}
+	var stdout, stderr io.Writer
+	if wait {
+		log := logger.Get(ctx)
+		stdout = log.Writer(logger.DebugLvl)
+		stderr = logger.NewFuncLogger(log.SupportsColor(), log.Level(),
+			func(level logger.Level, fields logger.Fields, b []byte) error {
+				// there's no kubeconfig in the container so k3d will emit confusing warnings
+				// note: no kubeconfig cleanup is necessary since k3d's execution is isolated
+				// 	via docker, so is never touching the host filesystem, but it's a weird
+				// 	use case so k3d doesn't have a flag to disable kubeconfig cleanup on delete
+				if bytes.Contains(b, []byte("Failed to remove cluster details")) ||
+					bytes.Contains(b, []byte("no such file or directory")) {
+					return nil
+				}
+				log.Write(logger.WarnLvl, b)
+				return nil
+			}).Writer(logger.WarnLvl)
+	}
+	if err := k.command(ctx, cmd, stdout, stderr, wait); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *K3dClient) GenerateKubeconfig(ctx context.Context, clusterName string) ([]byte, error) {
+	var kubeconfigBuf bytes.Buffer
+	stderr := logger.Get(ctx).Writer(logger.WarnLvl)
+	err := k.command(ctx, []string{"kubeconfig", "get", clusterName}, &kubeconfigBuf, stderr, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get kubeconfig: %v", err)
+	}
+	return kubeconfigBuf.Bytes(), nil
+}
+
+func (k *K3dClient) command(ctx context.Context, cmd []string, stdout io.Writer, stderr io.Writer, wait bool) error {
+	k.ensurePulled.Do(func() {
+		ref, err := k.cli.ImagePull(ctx, k.k3dImage)
+		if err != nil {
+			logger.Get(ctx).Errorf("failed to pull %q image: %v", k.k3dImage, err)
+		} else {
+			k.k3dImage = ref
+		}
+	})
+
+	runOpts := docker.RunOptions{
```

Comment:
is the idea of this to avoid tooling to install k3d?

I'm a little a bit about having k3d state living inside of a container and being hard to cleanup

Reply:
> is the idea of this to avoid tooling to install k3d?

Precisely - the value prop of "download tilt and run `tilt demo` to try it out" vanishes if you need to download k3d/minikube/KIND - at that point, you might as well actually create the cluster yourself 🙃 

(TBH I think that an approach either like this in Docker OR that actually downloads k3d/kind/minikube before execing if needed would probably be a good addition to `ctlptl`, and ideally we could use that here.)

> I'm a little a bit about having k3d state living inside of a container and being hard to cleanup

There's not any state other than the Docker containers it creates - each invocation of `k3d` actually runs in an ephemeral container and gets thrown away. Clean-up wise, I delegated to `k3d cluster delete` for ease, but it actually might be more sane to just use some labels and delete the containers directly based off that, which is presumably what it's doing anyway. (The generated `kubeconfig` is at a tmp path to not interfere with any the user might already have, so there's no config cleanup needed.)

LMK if that makes sense or if I'm overlooking something here!

Reply:
I went ahead and opened #4995 with the k3d-via-Docker client portion to (hopefully?) make it easier to discuss